torch/csrc/quantized/quantized_backward.cpp

```diff
@@ -0,0 +1,121 @@
+#include <ATen/native/quantized/PackedParams.h>
+#include <ATen/native/quantized/cpu/QnnpackUtils.h>
+#include <torch/csrc/autograd/FunctionsManual.h>
+#include <torch/csrc/autograd/VariableTypeUtils.h>
+#include <torch/csrc/autograd/functions/utils.h>
+#include <torch/custom_class.h>
+#include <torch/library.h>
+
+// This function is very specific to QNNPACK, if application does not use
+// QNNPACK, Gradient function dispatching does not need to be registered.
+
+#ifdef USE_PYTORCH_QNNPACK
+namespace caffe2 {
+CAFFE_KNOWN_TYPE(c10::intrusive_ptr<PackedLinearWeightsQnnp>);
+}
+namespace {
+using namespace torch::autograd;
+using namespace generated::details;
+using namespace at;
+// Gradient into argument 1 (`input`). TODO: Support bias as well.
+struct TORCH_API PackedLinearWeightDynamicBackward : public TraceableFunction {
```

Comment:
You shouldn't write custom Node by hand. Use a custom Function to tie a fw and bw together: https://pytorch.org/tutorials/advanced/cpp_autograd.html#using-custom-autograd-function-in-c

In this case, the fw should just be the redispatch call like you do below.

Reply:
where is the fw call btw? I only see bw implementation here.

Reply:
@albanD I'm honestly unsure if it is a feasible to implement this way, we lost an access to original weight with quantized tensor has a backend QNNPACK, it frees up weight tensor.
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp#L335

Reply:
> where is the fw call btw?

The fw is in the backend key. The Autograd key is only responsible for setting up the graph.

Reply:
> I'm honestly unsure if it is a feasible to implement this way, we lost an access to original weight tensor

I'm not sure what that changes? The signature of the custom Function is the same as packed_linear_weight_grad (you want to add the keyset to redispatch).
And from there you can save anything you need the same way you do with this one.